### PR TITLE
Filter 'format' #145

### PR DESCRIPTION
**Pull request by mateka**
First take-on #145. It isn't the most pretty, because format internals do not own arguments...

### CODE REVIEW

**Comment by flexferrum** [comment]
Good job, thanks! But according to this: https://travis-ci.org/jinja2cpp/Jinja2Cpp/jobs/597036948
There is memory leak in the implementation.

**Comment by flexferrum** [comment]
Also there is segfault in one testcase: 
```
1: unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.
1: [  FAILED  ] Format/FilterGenericTest.Test_Wide/7, where GetParam() = 'PI = {pi:.2f}' | format(pi=3.1415) | pprint -> 'PI = 3.14' (1 ms)
```

File: src/filters.cpp [hunk 1]
```diff
@@ -834,16 +834,190 @@ InternalValue Slice::Batch(const InternalValue& baseVal, RenderContext& context)
...
+
+class NamedArgumentCreator
+{
+public:
+    NamedArgumentCreator(const std::string& name, ValuesBuffer& valuesBuffer)
+        : m_name(name), m_valuesBuffer(valuesBuffer)
+    {}
+
+    template<typename T>
+    const auto& operator()(const T& t) const
```
Comment (flexferrum) [inline src/filters.cpp]: I suppose, the problem with `use-after-free` error is here. When you push into vector and reallocation occurred (which is possible even with reserve) all references to the previous items are invalidated.

Comment (mateka) [reply src/filters.cpp]: The funny thing is that, I cannot get sanitizer problems on my Ubuntu vm. VS2019 won't find any problems either. I believe that push_backs shouldn't do reallocation: https://en.cppreference.com/w/cpp/container/vector/push_back, https://stackoverflow.com/a/10735186. I will look into it a little more.

File: src/filters.cpp [hunk 2]
```diff
@@ -868,16 +868,190 @@ InternalValue Slice::Batch(const InternalValue& baseVal, RenderContext& context)
...
+    bool,
+    std::string,
+    int64_t,
+    double,
+    NamedArgument<bool>,
+    NamedArgument<std::string>,
+    NamedArgument<int64_t>,
+    NamedArgument<double>
+>;
+using ValuesBuffer = std::vector<ValueHandle>;
```
Comment (flexferrum) [inline src/filters.cpp]: What if you replace vector with deque here?

Comment (mateka) [reply src/filters.cpp]: Arghh! Github merged master into my branch, but it lost commit, which I wanted to try. I will try it again and, if it won't work, I will try deque.

**Comment by flexferrum** [comment]
It works! :) Can you now add extra tests to make codecov report green? https://codecov.io/gh/jinja2cpp/Jinja2Cpp/compare/4aece43f59f22ffe2d4bd4aa1c486caa03f17a28...d4e791b0abaa55583182768ba8ae7626ea749982/diff

**Comment by flexferrum** [comment]
Thanks a lot for the help!

**Pull request merged**